### PR TITLE
Autocomplete statements #1066

### PR DESCRIPTION
**Pull request by davidpofo**

### CODE REVIEW

File: controls/models.py [hunk 1]
```diff
@@ -61,6 +62,41 @@ def catalog_control_as_dict(self):
...
         return catalog_control_dict[self.sid]
 
+    def create_prototype(self):
+        """Creates a prototype statement from an existing statement and prototype object"""
+
+        if self.prototype is not None:
+            # Prototype already exists for statement
+            return self.prototype
+            # check if prototype content is the same, report error if not, or overwrite if permission approved
+        from copy import deepcopy
```
Comment (bama4) [inline controls/models.py:72]: Unless you are doing a conditional import, you may want to consider placing all imports at the top of the file

File: controls/models.py [hunk 2]
```diff
@@ -61,6 +62,41 @@ def catalog_control_as_dict(self):
...
+        return self.prototype
+
+    def create_instance_from_prototype(self, consumer_element_id):
+        """Creates a control_implementation statement instance for root_element of a system from an existing control implementation prototype statement"""
+
+        # if self.prototype is not None:
+        #     # Prototype already exists for statement
+        #     return self.prototype
+        #     # check if prototype content is the same, report error if not, or overwrite if permission approved
+        from copy import deepcopy
```
Comment (bama4) [inline controls/models.py:90]: Same here (see import comment)

**Review by gregelin — CHANGES_REQUESTED** [review]
Please do not make any changes to `autocomplete_statements_#1066`. I am in the process of synchronizing 3 branches: 
- autocomplete_statements_#1066
- ge/reuse-0903
- automated-tests-statements

**Comment by gregelin** [comment]
The following three branches are currently synchronized as of 18934669:

- `autocomplete_statements_#1066`
- `ge/reuse-0903`
- `automated-tests-statements`

**Comment by gregelin** [comment]
Still need to be addressed: Making sure solution works for statements associated with parts. 

Reproducing error:
1. Create a control statement for a component and associated statement with a control part
2. Create a second control statement for a component and associated statement with a different control part
3. Change the local control statement for one of the parts.
4. Update the certified statement for the changed control.
5. ERROR: the updated certified statement is updated for the control statements for both parts of the control.

What should happened is the statements for different control parts are treated differently.

File: CHANGELOG.md [hunk 3]
```diff
@@ -132,6 +165,7 @@ Enable upgrade of project root_task to more recent version.
 **Documentation changes**
 
 * Document new logger entries for successful and failed attempts to upgrade project's root_task app.
+>>>>>>> master
```
Comment (davidpofo) [inline CHANGELOG.md:168]: Left over from fixing conflicts I am guessing

File: controls/tests.py [hunk 4]
```diff
@@ -427,5 +304,127 @@ def test_element_create(self):
         # poam.delete()
         # self.assertTrue(poam.uuid is None)
 
+class ControlComponentTests(OrganizationSiteFunctionalTests):
 
+    def create_test_statement(self, sid, sid_class, body, statement_type, status):
```
Comment (davidpofo) [inline controls/tests.py:309]: This was actually from an early iteration it is not used in this set of tests.

**Review by omarabed15 — CHANGES_REQUESTED** [review]

File: controls/tests.py [hunk 5]
```diff
@@ -508,5 +429,127 @@ def test_element_create(self):
...
+        self._login()
+        self._new_project()
+        var_sleep(1)
+
+        # Baseline selection
+        self.navigateToPage("/systems/1/controls/selected")
+        # Select moderate
+        self.navigateToPage("/systems/1/controls/baseline/NIST_SP-800-53_rev4/moderate/_assign")
+        # Head to the control ac-3
+        self.navigateToPage("/systems/1/controls/catalogs/NIST_SP-800-53_rev4/control/ac-3")
```
[the remark this answers is not in the document]
Comment (davidpofo) [reply controls/tests.py]: We actually need to hit line 497 to assign a baseline to the system we are on(system 1). I'll remove 495 as its not needed thanks for pointing that out!

File: controls/utilities.py [hunk 6]
```diff
@@ -63,3 +63,18 @@ def oscalize_control_id(cl_id):
...
+    NIST_SP_800_53_rev4 --> NIST_SP-800-53_rev4
+    """
+
+    # A default catalog key
+    if catalogkey=='':
+        catalogkey = 'NIST_SP-800-53_rev4'
+    # Handle improperly formatted control id
+    if catalogkey.count("_") > 2:
+        split_key_list = catalogkey.split("_800_")
+        catalogkey = split_key_list[0] + "-800-" + split_key_list[1]
```
[the remark this answers is not in the document]
Comment (davidpofo) [reply controls/utilities.py:78]: No this is consistent as it refers to the sid class. Check out the class Catalogs.

File: controls/utilities.py [hunk 7]
```diff
@@ -63,3 +63,18 @@ def oscalize_control_id(cl_id):
...
 
+
+def oscalize_catalog_key(catalogkey):
+    """ Covers empty catalog key case. Otherwise, outputs an oscal standard catalog key from various common formats for catalog keys
+    NIST_SP_800_53_rev4 --> NIST_SP-800-53_rev4
+    """
+
+    # A default catalog key
+    if catalogkey=='':
+        catalogkey = 'NIST_SP-800-53_rev4'
```
[the remark this answers is not in the document]
Comment (davidpofo) [reply controls/utilities.py:74]: Not sure if that is necessary as this is already a utility function to provide a default catalog key when empty. No opinion on the default though just needed one of them for an empty edgecase.  @gregelin is this sid_class appropriate as a default?

**Pull request merged**